blog/_posts/2021-02-16-improving-scala-library-ecosystem.md

```diff
@@ -0,0 +1,127 @@
+---
+layout: blog-detail
+post-type: blog
+by: Eugene Yokota, Julien Richard-Foy
+title: Improving the Scala Library Ecosystem
+---
+
+One of the things that makes Scala powerful and fun to use is its library ecosystem — spanning across the Maven ecosystem for JVM, and npm for Scala.JS. We can use libraries to do operations like download files from the web, or we can use it to adopt different programming paradigms. The library ecosystem allows us to write more useful programs with less effort.
```

Comment:
Ooh, very subtle!  I happen to agree that "make" is preferable to "makes" in this case, but you can get whole pages written about (and vociferous disagreement about) which form is correct, or whether both are but have slightly different meanings.

I prefer the plural-agreement as suggested in the change because it more strongly suggests that there are other things that make Scala powerful and fun to use.  (Which is true, I think!)  Especially to programmers, who are liable to be more logical in how they assign scope.  (The number agreement of "make" suggests "One of the (things that make Scala powerful and fun to use) is (its library ecosystem)".)

Reply:
Ah ah! I didn't think this was even controversial. I thought this was simply a typo. How do you parse the sentence if it is singular on purpose? "One (of the things) (that makes Scala powerful and fun to use) is (its library ecosystem)" doesn't seem like a valid parse tree to me: "of the things" calls for something more.

Reply:
Ha, interesting! I don't think I've ever thought about this before.

Data point: no matter how illogical it might be, "makes" is definitely what would actually come out of my native-speaker mouth. It would never occur to me to say "make".

Reply:
English is full of weird "rules", including number-agreement-override in certain cases including, usually, this one.

The reason, inasmuch as there is one, to my understanding, is just drawing a parallel to other common constructs.

1. One dog makes a hole.
2. One dog that I love makes a hole.
3. One dog out of the pack makes a hole.
4. One of the dogs makes a hole.

After all that number agreement, how can you go against it?

The alternate parse is: "(One of the things) (_whose individual distinguishing characteristic is_ that _it_ makes Scala powerful and fun to use) is (its library ecosystem)".

This is an unlikely parse, because you have no context for pre-existing things, so adding a comma-free parenthetical about this particular thing is odd.

Anyway, human language is weird.  But it's not _forbidden_ to help push the meaning towards "One of (the Xs with quality Y) is Z" and away from "(One of the Xs) has (quality Y) and is Z" by agreeing with the plural, even if the convention is for that form to have number override even when you mean the former.

So the change is, I think, good, despite possibly feeling a little weird to some native speakers.